src/app/fdctl/run/tiles/fd_metric.c

```diff
@@ -0,0 +1,476 @@
+  fd_topo_t * topo;
+
+  int socket_fd;
+
+  fd_metric_connection_t conns[ MAX_CONNS ];
+  struct pollfd            fds[ MAX_CONNS+1 ];
+
+  ulong conn_id;
+} fd_metric_ctx_t;
+
+FD_FN_CONST static inline ulong
+scratch_align( void ) {
+  return 128UL;
+}
+
+FD_FN_PURE static inline ulong
+scratch_footprint( fd_topo_tile_t * tile ) {
+  (void)tile;
+  ulong l = FD_LAYOUT_INIT;
+  l = FD_LAYOUT_APPEND( l, alignof( fd_metric_ctx_t ), sizeof( fd_metric_ctx_t ) );
+  return FD_LAYOUT_FINI( l, scratch_align() );
+}
+
+FD_FN_CONST static inline void *
+mux_ctx( void * scratch ) {
+  return (void*)fd_ulong_align_up( (ulong)scratch, alignof( fd_metric_ctx_t ) );
+}
+
+static void
+close_conn( fd_metric_ctx_t * ctx,
+            ulong             idx ) {
+  if( FD_UNLIKELY( -1==close( ctx->fds[ idx ].fd ) ) ) FD_LOG_ERR(( "close failed (%i-%s)", errno, strerror( errno ) ));
+  ctx->fds[ idx ].fd = -1;
+}
+
+static void
+accept_conns( fd_metric_ctx_t * ctx ) {
+  for(;;) {
+    int fd = accept( ctx->socket_fd, NULL, NULL );
+
+    if( FD_UNLIKELY( -1==fd ) ) {
+      if( FD_LIKELY( EAGAIN==errno ) ) break;
+      else if( FD_LIKELY( ENETDOWN==errno || EPROTO==errno || ENOPROTOOPT==errno || EHOSTDOWN==errno ||
+                          ENONET==errno || EHOSTUNREACH==errno || EOPNOTSUPP==errno || ENETUNREACH==errno ) ) continue;
+      else FD_LOG_ERR(( "accept failed (%i-%s)", errno, strerror( errno ) ));
+    }
+
+    /* Just evict oldest connection if it's still alive, they were too slow. */
+    if( FD_UNLIKELY( -1!=ctx->fds[ ctx->conn_id ].fd ) ) close_conn( ctx, ctx->conn_id );
+
+    ctx->fds[ ctx->conn_id ].fd = fd;
+    ctx->conns[ ctx->conn_id ] = (fd_metric_connection_t){
+      .bytes_read    = 0UL,
+      .bytes_written = 0UL,
+      .output_len    = 0UL,
+    };
+    ctx->conn_id = (ctx->conn_id + 1) % MAX_CONNS;
+  }
+}
+
+#define PRINT( ... ) (__extension__({                  \
+    int n = snprintf( *out, *out_len, __VA_ARGS__ );   \
+    if( FD_UNLIKELY( n<0 ) ) return -1;                \
+    if( FD_UNLIKELY( (ulong)n>=*out_len ) ) return -1; \
+    *out += n; *out_len -= (ulong)n;                   \
+    n;                                                 \
+  }))
+
+#define PRINT_LINK_IN  (0)
+#define PRINT_LINK_OUT (1)
+#define PRINT_TILE     (2)
+
+static ulong
+find_producer_out_idx( fd_topo_t *      topo,
+                       fd_topo_tile_t * producer,
+                       fd_topo_tile_t * consumer,
+                       ulong            consumer_in_idx ) {
+  ulong count = 0UL;
+  for( ulong i=0; i<topo->tile_cnt; i++ ) {
+    fd_topo_tile_t * consumer_tile = &topo->tiles[ i ];
+    for( ulong j=0; j<consumer_tile->in_cnt; j++ ) {
+      if( FD_UNLIKELY( consumer_tile->in_link_id[ j ] == producer->out_link_id_primary && consumer_tile->in_link_reliable[ j ] ) ) {
+        if( FD_UNLIKELY( consumer==consumer_tile && consumer_in_idx==j ) ) return count;
+        count++;
+      }
+    }
+  }
+  return ULONG_MAX;
+}
```

Comment:
Given all the trouble we had with a similar-looking piece of code yesterday or the day before, this one could use some more comments. I think it finds all reliable consumers of producer's primary output and then gives you the position of consumer within that list, right?